File: src/main/java/au/id/micolous/metrodroid/transit/intercode/IntercodeLookupNavigo.java
/*
 * IntercodeLookupNavigo.java
 *
 * Copyright 2009-2013 by 'L1L1'
 * Copyright 2018 Google
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package au.id.micolous.metrodroid.transit.intercode;

import android.util.SparseArray;

import au.id.micolous.farebot.R;
import au.id.micolous.metrodroid.multi.Localizer;
import au.id.micolous.metrodroid.transit.Station;
import au.id.micolous.metrodroid.util.StationTableReader;
import au.id.micolous.metrodroid.util.Utils;

class IntercodeLookupNavigo extends IntercodeLookupSTR {
    private static final SparseArray<String> SECTOR_NAMES = new SparseArray<>();
    private static final String NAVIGO_STR = "navigo";

    static {
        // TODO: Move this to MdSt
        SECTOR_NAMES.put(1, "Cité");
        SECTOR_NAMES.put(2, "Rennes");
        SECTOR_NAMES.put(3, "Villette");
        SECTOR_NAMES.put(4, "Montparnasse");
        SECTOR_NAMES.put(5, "Nation");
        SECTOR_NAMES.put(6, "Saint-Lazare");
        SECTOR_NAMES.put(7, "Auteuil");
        SECTOR_NAMES.put(8, "République");
        SECTOR_NAMES.put(9, "Austerlitz");
        SECTOR_NAMES.put(10, "Invalides");
        SECTOR_NAMES.put(11, "Sentier");
        SECTOR_NAMES.put(12, "Île Saint-Louis");
        SECTOR_NAMES.put(13, "Daumesnil");
        SECTOR_NAMES.put(14, "Italie");
        SECTOR_NAMES.put(15, "Denfert");
        SECTOR_NAMES.put(16, "Félix Faure");
        SECTOR_NAMES.put(17, "Passy");
        SECTOR_NAMES.put(18, "Étoile");
        SECTOR_NAMES.put(19, "Clichy - Saint Ouen");
        SECTOR_NAMES.put(20, "Montmartre");
        SECTOR_NAMES.put(21, "Lafayette");
        SECTOR_NAMES.put(22, "Buttes Chaumont");
        SECTOR_NAMES.put(23, "Belleville");
        SECTOR_NAMES.put(24, "Père Lachaise");
        SECTOR_NAMES.put(25, "Charenton");
        SECTOR_NAMES.put(26, "Ivry - Villejuif");
        SECTOR_NAMES.put(27, "Vanves");
        SECTOR_NAMES.put(28, "Issy");
        SECTOR_NAMES.put(29, "Levallois");
        SECTOR_NAMES.put(30, "Péreire");
        SECTOR_NAMES.put(31, "Pigalle");
    }

    IntercodeLookupNavigo() {
        super(NAVIGO_STR);
    }

    private static final int RATP = 3;
    private static final int SNCF = 2;

    @Override
    public Station getStation(int locationId, Integer agency, Integer transport) {
        if (locationId == 0)
            return null;
        if (agency == null)
            agency = 0;
        if (transport == null)
            transport = 0;
        int mdstStationId = locationId | (agency << 16) | (transport << 24);
        int sector_id = locationId >> 9;
        int station_id = (locationId >> 4) & 0x1F;
        String humanReadableId = Integer.toString(locationId);
        String fallBackName = Integer.toString(locationId);
        if (transport == IntercodeTransaction.TRANSPORT_TRAIN && (agency == RATP || agency == SNCF)) {
            mdstStationId = (mdstStationId & 0xff00fff0) | 0x30000;
        }
        if ((agency == RATP || agency == SNCF)  && (transport == IntercodeTransaction.TRANSPORT_METRO || transport== IntercodeTransaction.TRANSPORT_TRAM)) {
             mdstStationId = (mdstStationId & 0x0000fff0) | 0x3020000;
             // TODO: i18n
             if (SECTOR_NAMES.get(sector_id) != null)
                 fallBackName = "sector " + SECTOR_NAMES.get(sector_id) + " station " + station_id;
             else
                 fallBackName = "sector " + sector_id + " station " + station_id;
             humanReadableId = sector_id + "/" + station_id;
        }

        Station st = StationTableReader.getStationNoFallback(NAVIGO_STR, mdstStationId, humanReadableId);
        if (st != null)
            return st;
        return Station.Companion.unknown(fallBackName);
    }

    @Override
    public String getSubscriptionName(Integer agency, Integer contractTariff) {
        if (contractTariff == null)
            return null;
        switch (contractTariff) {
            case 0:
                // TODO: i18n
                return "Forfait";
        }
        return Localizer.INSTANCE.localizeString(R.string.unknown_format, contractTariff);
    }
}

File: src/commonMain/kotlin/au/id/micolous/metrodroid/transit/intercode/IntercodeLookupNavigo.kt
/*
 * IntercodeLookupNavigo.kt
 *
 * Copyright 2009-2013 by 'L1L1'
 * Copyright 2018 Google
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package au.id.micolous.metrodroid.transit.intercode

import au.id.micolous.metrodroid.card.CardType
import au.id.micolous.metrodroid.multi.Localizer
import au.id.micolous.metrodroid.multi.R
import au.id.micolous.metrodroid.multi.StringResource
import au.id.micolous.metrodroid.transit.CardInfo
import au.id.micolous.metrodroid.transit.Station
import au.id.micolous.metrodroid.transit.en1545.En1545Parsed
import au.id.micolous.metrodroid.transit.en1545.En1545Transaction
import au.id.micolous.metrodroid.transit.en1545.En1545TransitData
import au.id.micolous.metrodroid.transit.TransitRegion
import au.id.micolous.metrodroid.util.StationTableReader

private const val NAVIGO_STR = "navigo"

internal object IntercodeLookupNavigo : IntercodeLookupSTR(NAVIGO_STR) {
    override fun cardInfo(env: () -> En1545Parsed): CardInfo =
            if (env().getIntOrZero(En1545TransitData.HOLDER_CARD_TYPE) == 1) NAVIGO_DECOUVERTE_CARD_INFO else NAVIGO_CARD_INFO

    override val allCards: List<CardInfo>
        get() = listOf(NAVIGO_CARD_INFO)

    override fun getStation(station: Int, agency: Int?, transport: Int?): Station? {
        if (station == 0)
            return null
        var mdstStationId = station or ((agency ?: 0) shl 16) or ((transport ?: 0) shl 24)
        val sectorId = station shr 9
        val stationId = station shr 4 and 0x1F
        var humanReadableId = station.toString()
        var fallBackName = station.toString()
        if (transport == En1545Transaction.TRANSPORT_TRAIN && (agency == RATP || agency == SNCF)) {
            mdstStationId = mdstStationId and -0xff0010 or 0x30000
        }
        if ((agency == RATP || agency == SNCF) && (transport == En1545Transaction.TRANSPORT_METRO || transport == En1545Transaction.TRANSPORT_TRAM)) {
            mdstStationId = mdstStationId and 0x0000fff0 or 0x3020000
            fallBackName = if (SECTOR_NAMES[sectorId] != null)
                Localizer.localizeString(R.string.navigo_sector_station_id,
                        SECTOR_NAMES[sectorId], stationId)
            else
                Localizer.localizeString(R.string.navigo_sector_id_station_id,
                        sectorId, stationId)
            humanReadableId = "$sectorId/$stationId"
        }

        return StationTableReader.getStationNoFallback(NAVIGO_STR, mdstStationId, humanReadableId)
                ?: Station.unknown(fallBackName)
    }


    override val subscriptionMap: Map<Int, StringResource> = mapOf(
            0 to R.string.navigo_forfait,
            3 to R.string.navigo_forfait_jour
    )

    private val SECTOR_NAMES = mapOf(
            // TODO: Move this to MdSt
            1 to "Cité",
            2 to "Rennes",
            3 to "Villette",
            4 to "Montparnasse",
            5 to "Nation",
            6 to "Saint-Lazare",
            7 to "Auteuil",
            8 to "République",
            9 to "Austerlitz",
            10 to "Invalides",
            11 to "Sentier",
            12 to "Île Saint-Louis",
            13 to "Daumesnil",
            14 to "Italie",
            15 to "Denfert",
            16 to "Félix Faure",
            17 to "Passy",
            18 to "Étoile",
            19 to "Clichy - Saint Ouen",
            20 to "Montmartre",
            21 to "Lafayette",
            22 to "Buttes Chaumont",
            23 to "Belleville",
            24 to "Père Lachaise",
            25 to "Charenton",
            26 to "Ivry - Villejuif",
            27 to "Vanves",
            28 to "Issy",
            29 to "Levallois",
            30 to "Péreire",
            31 to "Pigalle"
    )

    private const val RATP = 3
    private const val SNCF = 2

    private val NAVIGO_CARD_INFO = CardInfo(
            name = "Navigo",  // personalised card
            imageId = R.drawable.navigo,
            imageAlphaId = R.drawable.iso7810_id1_alpha,
            locationId = R.string.location_paris,
            region = TransitRegion.FRANCE,
            cardType = CardType.ISO7816)

    private val NAVIGO_DECOUVERTE_CARD_INFO = CardInfo(
            name = "Navigo découverte",  // anonymous card (lit: discovery)
            imageId = R.drawable.navigo,
            imageAlphaId = R.drawable.iso7810_id1_alpha,
            locationId = R.string.location_paris,
            region = TransitRegion.FRANCE,
            cardType = CardType.ISO7816)
}
